api/src/poem_backend/response.rs

```diff
@@ -95,105 +108,261 @@ pub enum AptosResponseContent<T: ToJSON + Send + Sync> {
+            }
+        }
+
+        // Generate a From impl that builds a response from a Bcs<Vec<u8>> and friends.
+        impl<T: poem_openapi::types::ToJSON + Send + Sync> From<($crate::poem_backend::bcs_payload::Bcs<Vec<u8>>, &aptos_api_types::LedgerInfo, [<$enum_name Status>])>
+            for $enum_name<T>
+        {
+            fn from(
+                (value, ledger_info, status): (
+                    $crate::poem_backend::bcs_payload::Bcs<Vec<u8>>,
+                    &aptos_api_types::LedgerInfo,
+                    [<$enum_name Status>]
+                ),
+            ) -> Self {
+                let content = $crate::poem_backend::AptosResponseContent::Bcs(value);
+                Self::from((content, ledger_info, status))
+            }
+        }
+
+        // Generate a TryFrom impl that builds a response from a T, an AcceptType,
+        // and all the other usual suspects. It expects to be called with a generic
+        // parameter E: InternalError, with which we can build an internal error
+        // response in case the BCS serialization fails.
+        impl<T: poem_openapi::types::ToJSON + Send + Sync + serde::Serialize> $enum_name<T> {
+            pub fn try_from_rust_value<E: InternalError>(
+                (value, ledger_info, status, accept_type): (
+                    T,
+                    &aptos_api_types::LedgerInfo,
+                    [<$enum_name Status>],
+                    &$crate::poem_backend::AcceptType
+                ),
+            ) -> Result<Self, E> {
+                match accept_type {
+                    AcceptType::Bcs => Ok(Self::from((
+                        $crate::poem_backend::bcs_payload::Bcs(
+                            bcs::to_bytes(&value)
+                                .map_err(|e| E::internal(e.into()).error_code($crate::poem_backend::AptosErrorCode::BcsSerializationError))?
+                        ),
+                        ledger_info,
+                        status
+                    ))),
+                    AcceptType::Json => Ok(Self::from((
+                        poem_openapi::payload::Json(value),
+                        ledger_info,
+                        status
+                    ))),
+                }
+            }
+        }
+        }
+    };
 }
+
+// Generate a success response that only has an option for 200.
+generate_success_response!(BasicResponse, (200, Ok));
+
+// Generate traits defining a "from" function for each of these status types.
+// The error response then impls these traits for each status type they mention.
+generate_error_traits!(Internal, BadRequest, NotFound);
+
+// Generate an error response that only has options for 400 and 500.
+generate_error_response!(BasicError, (400, BadRequest), (500, Internal));
+
+// This type just simplifies using BasicResponse and BasicError together.
+pub type BasicResult<T> = poem::Result<BasicResponse<T>, BasicError>;
+
+// As above but with 404.
```

Comment:
We've got a ton of macros, this isn't the easiest to read, but let's not index too much on these.

Reply:
Agreed I'd like to come back to this, I'd like to find a non-macro approach here. Once you look past the macro complexity though the rest of the code becomes very nice and typed.